exercises/hangman/src/test/java/HangmanTest.java

```diff
@@ -0,0 +1,200 @@
+        assertEquals(Collections.singletonList(Part.HEAD), last.parts);
+        assertEquals(Status.PLAYING, last.status);
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void gameInProgress() {
+        Observable<Output> result = hangman.play(
+            Observable.fromArray("secret"),
+            Observable.fromArray("a", "e", "o", "s"));
+        Output last = result.blockingLast();
+        assertEquals("se__e_", last.discovered);
+        assertEquals(setOf("e", "s"), last.guess);
+        assertEquals(setOf("a", "o"), last.misses);
+        assertEquals(
+            Arrays.asList(
+                Part.HEAD,
+                Part.BODY),
+            last.parts);
+        assertEquals(Status.PLAYING, last.status);
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void wonGame() {
+        Observable<Output> result = hangman.play(
+            Observable.fromArray("secret"),
+            Observable.fromArray("a", "e", "o", "s", "c", "r", "g", "t"));
+        Output last = result.blockingLast();
+        assertEquals("secret", last.discovered);
+        assertEquals(setOf("c", "e", "r", "s", "t"), last.guess);
+        assertEquals(Status.WIN, last.status);
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void lostGame() {
+        Observable<Output> result = hangman.play(
+            Observable.fromArray("secret"),
+            Observable.fromArray("a", "b", "c", "d", "e", "f", "g", "h"));
+        Output last = result.blockingLast();
+        assertEquals("_ec_e_", last.discovered);
+        assertEquals(setOf("a", "b", "d", "f", "g", "h"), last.misses);
+        assertEquals(Status.LOSS, last.status);
+        assertEquals(
+            Arrays.asList(
+                Part.HEAD,
+                Part.BODY,
+                Part.LEFT_ARM,
+                Part.RIGHT_ARM,
+                Part.LEFT_LEG,
+                Part.RIGHT_LEG),
+            last.parts);
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void consecutiveGames() {
+        ObservableEmitter<String>[] emitters = new ObservableEmitter[2];
+        Runnable emit = () -> {
+            emitters[0].onNext("secret");
+            Stream.of("a", "e", "o", "s", "c", "r", "g", "t").forEach(emitters[1]::onNext);
+            emitters[0].onNext("abba");
+            Stream.of("a", "e", "s", "b").forEach(emitters[1]::onNext);
+            emitters[0].onComplete();
+        };
+        Observable<String> words = Observable.create(
+            emitter -> {
+                emitters[0] = emitter;
+                if (emitters[1] != null) {
+                    emit.run();
+                }
+            });
+        Observable<String> letters = Observable.create(
+            emitter -> {
+                emitters[1] = emitter;
+                if (emitters[0] != null) {
+                    emit.run();
+                }
+            });
+        Observable<Output> outputs = hangman.play(words, letters);
```

Comment:
This is tricky to read, can you add some descriptive some comments... If this test fails, the student will have to look at the test to try to figure out what it is doing.

Reply:
You are right. I think I will decompose with additional methods and add comments.